appinventor/blocklyeditor/src/blocks/text.js

```diff
@@ -32,18 +37,47 @@ Blockly.Blocks['text'] = {
 Blockly.Blocks.text.connectionCheck = function (myConnection, otherConnection) {
   var block = myConnection.sourceBlock_;
   var otherTypeArray = otherConnection.check_;
+  var textIsNumber = !isNaN(parseFloat(block.getFieldValue('TEXT')));
+
   for (var i = 0; i < otherTypeArray.length; i++) {
-    if (otherTypeArray[i] == "String") {
-      return true;
-    } else if (otherTypeArray[i] == "Number" && !isNaN(parseFloat(block.getFieldValue('TEXT')))) {
-      return true;
-    } else if (otherTypeArray[i] == "Key") {
+    var type = otherTypeArray[i];
+    if (type == "String" || type == "Key" || (type == "Number" && textIsNumber)) {
       return true;
     }
   }
   return false;
 };
 
+/**
+ * Bumps the text block out of its connection iff it is connected to a number
+ * input and it no longer contains a number.
+ * @param {string} finalValue The final value typed into the text input.
+ * @this Blockly.Block
+ */
+Blockly.Blocks.text.bumpBlockOnFinishEdit = function(finalValue) {
+  var connection = this.outputConnection.targetConnection;
+  if (!connection) {
+    return;
+  }
+  if (!isNaN(parseFloat(finalValue))) {
+    // Block is a number, so no matter where it lives it is valid.
+    return;
+  }
+
+  var typeArray = connection.check_;
```

Comment:
At this point, you could call `Blockly.Blocks.text.connectionCheck(this.outputConnection, this.outputConnection.otherConnection)` and check the return value. That way if we change the logic there we don't have to also change it here.

Reply:
That's a really good point! Will fix

Reply:
I'm going to need to move some stuff around to get this to work. Currently connectionCheck uses block.getFieldValue() to check if the field contains a number. But at the time onFinishEditing_ is called the fields value hasn't been updated yet :/ so it does still contain a number.

@ewpatton Is it ok if I wait until your hotfix is in to move stuff around? I think that'll be easier than trying to pull down your branch and rebase onto it.

Reply:
You could merge that branch into yours and write your code assuming it gets merged.

Reply:
What if we add an `opt_value` parameter to the other and if it's undefined we populate it with the block's value?

Reply:
sounds good to me!